databricks_cli/tunnel/api.py

```diff
@@ -0,0 +1,236 @@
+        private_key_path.parent.mkdir(parents=True, exist_ok=True)
+
+        if private_key_path.exists():
+            private_key_path.unlink()
+        private_key_path.write_bytes(private_key)
+        os.chmod(private_key_path, 0o600)
+
+        public_key_path = Path(f"{self.local_public_key_path}").expanduser()
+        if public_key_path.exists():
+            public_key_path.unlink()
+        public_key_path.write_bytes(public_key)
+        os.chmod(public_key_path, 0o600)
+
+    def check_remote_tunnel(self):
+        print("Ping the remote tunnel server...")
+        resp = requests.get(f"{self.tunnel_config.healthcheck_url}",
+                            headers=self._api_client.default_headers)
+        if resp.status_code != 200:
+            raise RuntimeError(f"Unable to connect to the remote tunnel server: {resp.status_code}")
+        status = resp.json()["status"]
+        if status != "working":
+            raise RuntimeError(f"Unexpected remote tunnel server status: {status}")
+
+        return True
+
+    def cleanup(self):
+        self.destroy_context()
+        self._init_params()
+
+    def start_tunneling(self, cluster_id, local_port=None):
+        self.cluster_id = cluster_id
+
+        try:
+            # check prerequisites:
+            # - running cluster
+            # - tunnel server is alive
+            self.check_cluster()
+            self.org_id = self.get_org_id()
+            self.tunnel_config = TunnelConfig(host=self._api_client.host,
+                                              token=self._api_client.token,
+                                              org_id=self.org_id,
+                                              cluster_id=self.cluster_id)
+            self.check_remote_tunnel()
+            # setup prerequisites:
+            # - exec context
+            # - setup ssh keys
+            if self.context_id is not None:
+                self.destroy_context()
+            self.context_id = self.create_context()
+            self.setup_ssh_keys()
+
+            self.start_local_server(local_port)
+        finally:
+            self.cleanup()
+
+    def start_local_server(self, local_port=None):
+        server = StreamingServer(self.tunnel_config, debug=self.debug)
+        if local_port:
+            server.listen(local_port)
+        else:
+            sockets = tornado.netutil.bind_sockets(0, '')
+            if sockets:
+                server.add_sockets(sockets)
+                for s in sockets:
+                    local_port = s.getsockname()[1]
+                    if local_port is not None:
+                        break
+            if local_port is None:
+                raise RuntimeError("Could not find a free port")
+        print(f"Starting local tunneling on localhost:{local_port}")
+        print("Please use the following command for SSH connection:")
+        print(f"ssh root@localhost -p {local_port} -i {self.local_private_key_path}")
```

Comment:
Random, but I wonder whether we should use ANSI to print this line red or some other way to make it stand out.

Reply:
yeah, i was thinking of prettifying the logs... let's file followup tickets

Reply:
click.echo can help with this IIRC,

Reply:
The output generated by `click.echo` is still similar as `print` actually, cos we haven't really added the ansi coloring. We'll file follow up tickets for this